tensorflow_federated/python/core/impl/executors/federating_executor.py

```diff
@@ -664,35 +572,14 @@ async def federated_reduce(self, arg):
           'Expected 3 elements in the `federated_reduce()` argument tuple, '
           'found {}.'.format(len(arg.internal_representation)))
 
-    # Encrypt client tensors
-    pk_a, sk_a = await self._trusted_aggregator_generate_keys()
-    enc_clients_arg = await self._encrypt_client_tensors(arg, pk_a)
-    val_type = enc_clients_arg.type_signature
-    val = enc_clients_arg.internal_representation
-    # Store original client dtype before encryption for
-    # future decryption
-    orig_clients_dtype = arg.type_signature[0].member.dtype
+    # Setup channel keys
+    await self.channel.setup()
```

Comment:
`setup` should only be called once, and only invoked from eg inside `_move`

Reply:
shouldn't it be called during the instantiation of the `Executor`/`Strategy`? or does the setup phase need to be performed before each `_move`? maybe you mean that the results of `setup` should only be used during `_move`, which I agree with

Reply:
I would also second that the executor/strategy (and all of their methods) should not need to worry about channels, the only way to interact with those should be through the `_move` method

Reply:
good catch! I moved it to the `_move`. However in `setup` I added a flag to check if the channel has already been setup (key generated). If True then do nothing. It seems to be easier this way. `setup` uses async/await would be hard to call it in the constructor of `TrustedAggregatorIntrinsicStrategy`. Also `self._target_executors` would be missing (being created in `FederatingExecutor`).